pilot/pkg/status/reporter.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"k8s.io/client-go/rest"
+
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/utils/clock"
+
+	"istio.io/istio/pilot/pkg/model"
+)
+
+func NewIstioContext(stop <-chan struct{}) context.Context {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-stop
+		cancel()
+	}()
+	return ctx
+}
+
+type Reporter struct {
+	mu                  sync.RWMutex
+	status              map[string]string
+	reverseStatus       map[string][]string
+	dirty               bool
+	inProgressResources []Resource
+	client              v1.ConfigMapInterface
+	cm                  *corev1.ConfigMap
+	UpdateInterval      time.Duration
+	PodName             string
+	clock               clock.Clock
+	store               model.ConfigStore
+}
+
+// Starts the reporter, which watches dataplane ack's and resource changes so that it can update status leader
+// with distribution information
+func (r *Reporter) Start(restConfig *rest.Config, namespace string, stop <-chan struct{}) {
+	ctx := NewIstioContext(stop)
+	t := r.clock.Tick(r.UpdateInterval)
+	clientSet, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		scope.Fatalf("Could not connect to kubernetes: %s", err)
+	}
+	r.client = clientSet.CoreV1().ConfigMaps(namespace)
+	r.cm = &corev1.ConfigMap{
+		ObjectMeta: v12.ObjectMeta{
+			Name:   (r.PodName + "_DistRpt"),
```

Comment:
does this mean we are going to have a bunch of config maps (100s, possibly)? Why not just a single one where they all enter their own keys or something? Are they cleaned up when a pod dies?

Reply:
Also `_DistRpt` is probably not a good suffix, mostly because its not a valid name :slightly_smiling_face:

Reply:
If one large configmap was used, kubernetes concurrency model would only allow one concurrent write to succeed, with others failing on the etag. This would result in retires, which increases load on the API.  Cleanup is a best-effort affair at line 80.